idaes/core/initialization/general_hierarchical.py

```diff
@@ -0,0 +1,156 @@
+
+        Returns:
+            Pyomo solver results object
+        """
+        if not hasattr(model, "control_volume"):
+            raise TypeError(
+                f"Model {model.name} does not appear to be a standard form unit model. "
+                f"Please use an Initializer specific to the model being initialized."
+            )
+
+        # Get logger
+        _log = self.get_logger(model)
+
+        sub_initializers = {}
+        for sm in model._initialization_order:
+            if sm is not model:
+                # Get initializers for add-ons
+                sub_initializers[sm] = self.get_submodel_initializer(sm)
+
+                # Call prepare method for add-ons
+                # TODO: Need arguments for submodel initialization
+                sub_initializers[sm].prepare()
+
+        # Initialize properties
+        try:
+            # Guess a 0-D control volume
+            # Initialize inlet properties - inlet state should already be fixed
+            prop_init = self.get_submodel_initializer(
+                model.control_volume.properties_in
+            )
+
+            if prop_init is not None:
+                prop_init.initialize(
+                    solver=self.config.solver,
+                    optarg=self.config.solver_options,
+                    outlvl=self.config.output_level,
+                )
+
+            # Map solution from inlet properties to outlet properties
+            state = to_json(
+                model.control_volume.properties_in,
+                wts=StoreSpec().value(),
+                return_dict=True,
+            )
+            from_json(
+                model.control_volume.properties_out,
+                sd=state,
+                wts=StoreSpec().value(only_not_fixed=True),
+            )
+        except AttributeError:
+            # Assume it must be a 1-D control volume
+            # TODO: Add steps here
+            raise
+
+        _log.info_high("Step 1: properties initialization complete")
+
+        # Initialize reactions if they exist
+        try:
+            model.control_volume.reactions.initialize(
+                solver=self.config.solver,
+                optarg=self.config.solver_options,
+                outlvl=self.config.output_level,
+                state_vars_fixed=True,
+            )
+        except AttributeError:
+            pass
+        _log.info_high("Step 2: reactions initialization complete")
+
+        # Solve full unit model
+        solver = get_solver(self.config.solver, self.config.solver_options)
+        solve_log = idaeslog.getSolveLogger(
+            model.name, self.config.output_level, tag="unit"
+        )
+
+        with idaeslog.solver_log(solve_log, idaeslog.DEBUG) as slc:
+            results = solver.solve(model, tee=slc.tee)
+
+        _log.info_high("Step 3: model {}.".format(idaeslog.condition(results)))
+
+        if len(model._initialization_order) > 1:
+            # Initialize add-ons
```

Comment:
What do we mean by `add_ons`?

Reply:
Things like costing blocks, where there is a sub-model added-on to the main model outside of the `build` method. In these cases, the main model has no idea of what is in the sub-model or how to initialize it itself, so it needs to call out to a separate initializer for the add-on.